### PR TITLE
feature(component): media gallery

### PR DESCRIPTION
**Pull request by gepifanio**
Ticket: https://github.com/unimelb/pattern-lib/issues/1265
*The way that the video is getting paused, I reckon could be improved.

- [x] Improve the class names to more BEM like

https://deploy-preview-1215--pattern-lib-unimelb.netlify.com/?selectedKind=Media%20Gallery

### CODE REVIEW

**Comment by gepifanio** [comment]
Comment: Deploy Started

Building with commit 615ae7179f5ecd36f6a3d25a70cf3bc8b289d873

https://app.netlify.com/sites/pattern-lib-unimelb/deploys/5d3927f63ff14400089e7449

**Comment by gepifanio** [comment]
@jonathan-doc considerations:

Gallery Changes (201907251219)

General

- [ ]	Set active thumbnail to have a border : 2px solid white
- [ ] Set overlay / in-page description text colour to Black Mid 1 #C6C8CA, + font-size to 0.9375rem, + line-height to 1.5625
- [ ] Make featured images unclickable for now (or until the interaction opens the full image in a new tab or something like that)
- [ ] Video should stop / pause on navigation
- [ ] Increase next / previous button padding to 2rem
- [ ] On Desktop, set the image title and description text alignment to center (since we’re presumably not including the fullscreen / download / share buttons, and they get lost to the left on wide displays)
- [ ] If the width of the thumbnail carousel exceeds the width of the viewport, it would be good to have the carousel scroll such that the active thumbnail is positioned at its centre. This would help users orient themselves, especially at smaller breakpoints where there are no next / previous controls visible


Overlay

- [ ] Set a quick fade transition for opening / closing the overlay (0.1s); it’s a bit jarring atm
- [ ] Investigate Image 7 and 9 – they seem to be affecting the height of the overlay components


Pop Over Thumbnails / Captions

- [ ] Don’t make caption headings clickable (alternatively, make them bring up the overlay if they need to be clickable)
- [ ] Increase top margin of caption heading to 1rem
- [ ] Set heading font-size to 1rem
- [ ] Remove heading italics
- [ ] Set description font-style to italic
- [ ] Set description colour to #575756
- [ ] Set description font-size to 0.875rem
- [ ] Centre-align headings & descriptions
- [ ] On Desktop, set the thumbnail container heights so that headings aren’t misaligned; maybe set the thumbnail containers to 250px (15.625rem). Conversely, Mobile shouldn’t have uniform heights for the thumbnail containers since there would only be a single column and we don’t want any extra whitespace


Standalone Thumbnails

- [ ] Don’t make thumbnails clickable
- [ ] Same caption styling as above

**Comment by gepifanio** [comment]
Hey @anormand,

Could you have a look at the `Media Gallery` component regarding accessibility?

https://deploy-preview-1215--pattern-lib-unimelb.netlify.com/?selectedKind=Media%20Gallery&selectedStory=In%20page&full=0&addons=1&stories=1&panelRight=0&addonPanel=REACT_STORYBOOK%2Freadme%2Fpanel

File: components/media-gallery/_media-gallery-overlay.css [hunk 1]
```diff
@@ -0,0 +1,76 @@
+.media-gallery--overlay {
+  display: none;
+  background: rgba(0, 0, 0, .9);
+  position: relative;
+  z-index: 3;
```
Comment (TimMansell) [inline components/media-gallery/_media-gallery-overlay.css]: we can use our z-index variables

File: components/media-gallery/_media-gallery.css [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    &--thumb {
+      cursor: pointer;
+      margin-right: .313rem;
+      opacity: .7;
+      height: 100%;
+
+      &.active {
+        opacity: 1;
+        fill: rgba(153, 162, 170, .25);
```
Comment (TimMansell) [inline components/media-gallery/_media-gallery.css]: --col-black-mid-2 for 153, 162, 170

File: components/media-gallery/_media-gallery.css [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        }
+      }
+
+      &--cover {
+        width: 100%;
+        height: 100%;
+        position: absolute;
+        top: 0;
+        left: 0;
+        z-index: 2;
```
Comment (TimMansell) [inline components/media-gallery/_media-gallery.css]: can use z-index variables

File: components/media-gallery/_thumbnails.css [hunk 4]
```diff
@@ -0,0 +1,85 @@
+@import '../_vars.css';
```
Comment (TimMansell) [inline components/media-gallery/_thumbnails.css]: Don't think we need this as it's imported globally.

File: components/media-gallery/_thumbnails.css [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    display: flex;
+    align-items: center;
+
+    ::before {
+      width: 100%;
+      height: 100%;
+      position: relative;
+      top: 0;
+      left: 0;
+      z-index: 2;
```
Comment (TimMansell) [inline components/media-gallery/_thumbnails.css]: can use z-index variables

File: components/media-gallery/_thumbnails.css [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    @media (--bp-desktop) {
+      height: 15.625rem;
+    }
+  }
+
+  &__description {
+    font-size: .875rem;
+    text-align: center;
+    font-style: italic;
+    color: #575756;
```
Comment (TimMansell) [inline components/media-gallery/_thumbnails.css]: Create a variable for this, or see if their is a colour close to it.  @jaredh-um

File: components/media-gallery/_media-gallery.css [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+  &__chevron {
+    color: #fff;
+    cursor: pointer;
+  }
+
+  &__container {
+    display: flex;
+    justify-content: center;
+    width: 100%;
+    height: 400px;
```
Comment (TimMansell) [inline components/media-gallery/_media-gallery.css]: convert to rem.

File: components/media-gallery/_media-gallery.css [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    margin: auto;
+    background: rgb(var(--col-black-dark));
+
+    @media (--bp-desktop) {
+      width: 80%;
+      height: 600px;
```
Comment (TimMansell) [inline components/media-gallery/_media-gallery.css]: convert to rem.

File: components/media-gallery/_media-gallery.css [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    }
+  }
+
+  &__thumbnails {
+    display: flex;
+    height: 74px;
```
Comment (TimMansell) [inline components/media-gallery/_media-gallery.css]: convert to rem.

File: components/media-gallery/_media-gallery.css [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+        opacity: 1;
+        fill: rgba(153, 162, 170, .25);
+        border: .125rem solid rgb(var(--col-background-primary-white));
+      }
+    }
+
+    .embed {
+      &--video {
+        position: relative;
+        width: 100px;
```
Comment (TimMansell) [inline components/media-gallery/_media-gallery.css]: convert to rem.

File: components/media-gallery/_thumbnails.css (hunk 6 above)
Comment (gepifanio) [reply components/media-gallery/_thumbnails.css]: Created a new variable called: `--col-grey-dark` and to be easier the value is #575757 or rbg(87, 87, 87);

**Comment by TimMansell** [comment]
- [ ] Bug with overlay 1 column.  Video doesn't show

**Comment by TimMansell** [comment]
- [ ] If caption goes onto two lines it moves the position of the close (x)

Screen Shot 2019-08-16 at 3 29 31 pm

**Comment by TimMansell** [comment]
- [ ] mobile viewport swiping or clicking on thumbnail image doesn't scroll thumbnail images.

File: components/media-gallery/MediaGallery.vue [hunk 11]
```diff
@@ -0,0 +1,282 @@
+<template>
+  <div>
+    <div :class="classes">
```
Comment (anormand) [inline components/media-gallery/MediaGallery.vue]: **Problem**
- Currently users of assistive technologies are not notified that they are entering a media gallery

**Solution**
- Add a role and aria-roledescription to the div

**Example**
Change
`<div class="media-gallery">`

to
`<div class="media-gallery" role="region" aria-roledescription="Media gallery">`

**Comment by gepifanio** [comment]
> * [x]  Bug with overlay 1 column.  Video doesn't show

Thumbnails are supposed to be 2, 3 or 4 column sizing and follow the standard grid rules.

**Comment by gepifanio** [comment]
https://deploy-preview-1215--pattern-lib-unimelb.netlify.com/?selectedKind=Media%20Gallery&selectedStory=In%20page&full=0&addons=1&stories=1&panelRight=0&addonPanel=REACT_STORYBOOK%2Freadme%2Fpanel

**Comment by gepifanio** [comment]
> * [ ]  mobile viewport swiping or clicking on thumbnail image doesn't scroll thumbnail images.

I have tried to fix that, the function `scrollIntoView` should be compatible with mobile devices, however, is not working in the small viewport, I couldn't figure out why.

**Comment by TimMansell** [comment]
- [ ] Fix image sizing in thumbnails.  Make same height

**Comment by TimMansell** [comment]
Hi @jonathan-doc, can you have a check.  It's pretty much there.  I have a few bugs to fix but design should be pretty close.  Main one is at the moment thumbnails images are not same height (to fix)

**Comment by TimMansell** [comment]
Hi @jonathan-doc, can test this component when you have some time.

**Comment by jonathan-doc** [comment]
@TimMansell Looks good, just found this:

- [ ] Set the spacing between the featured image and thumbnails to be the same as the spacing between individual thumbnails; it currently shrinks for viewport widths less than 900.
gallery-thumbnail-spacing

The featured image indicator shifts into view as needed, but only when using the arrow controls or keyboard for navigation; this is a problem on Mobile, where there are no arrow controls and the thumbnail carousel is not currently scrollable.

- [ ] Set the featured image indicator to shift into view for all other image navigation interactions (swipe, thumbnail press)
- [ ] Make the thumbnail carousel scrollable, so that users aren't forced to swipe through every image in the set

**Comment by TimMansell** [comment]
- [ ] Win10 IE11 thumbnails broken in inpage nav
- [ ] Fix scrollbar for windows browsers

**Comment by gepifanio** [comment]
the problem with the WIN10 IE11 has been solved but introduced a new smaller issue which is:

- [ ] Thumbnails are overlapping

image

**Comment by TimMansell** [comment]
Hi @jonathan-doc, all fixes have been done.  Can you do another quick test and accept PR if you're happy for us to merge.

**Comment by jonathan-doc** [comment]
Hey @TimMansell, just need:

- [ ] Show featured media indicator outline on video thumbnail

Also, we have a similar problem that the carousel had, in that the height of the title/description will be pushing page content dynamically. Are we able to set the height of the bottom matter (title, description) based on the lengthiest content in the set?

**Pull request merged**